diff --git a/src/GNUmakefile b/src/GNUmakefile
index 92a4935..8b1355b 100644
--- a/src/GNUmakefile
+++ b/src/GNUmakefile
@@ -13,6 +13,7 @@ LD       = $(CC)
 STRIP    ?= $(CROSS_COMPILE)strip
 OBJCOPY  ?= $(CROSS_COMPILE)objcopy
 OBJDUMP  ?= $(CROSS_COMPILE)objdump
+READELF  ?= $(CROSS_COMPILE)readelf
 
 CPPFLAGS += -D_FILE_OFFSET_BITS=64 -D_GNU_SOURCE -I. -I$(VPATH)
 CFLAGS   += -Wall -Wextra -O2
@@ -235,7 +236,7 @@ $(eval $(call build_loader,-m32))
 endif
 
 loader/loader-info.c: loader/loader
-	readelf -s $< | awk -f loader/loader-info.awk > $@
+	$(READELF) -s $< | awk -f loader/loader-info.awk > $@
 
 ######################################################################
 # Dependencies
diff --git a/src/cli/cli.c b/src/cli/cli.c
index bbc6400..832b54c 100644
--- a/src/cli/cli.c
+++ b/src/cli/cli.c
@@ -47,6 +47,12 @@
 #include <extension/extension.h>
 #include <extension/sysvipc/sysvipc.h>
 
+#ifdef __OHOS__
+#include "queue.h"    /* CIRCLEQ_*, */ // which is not present in ohos header
+#else
+#include <sys/queue.h> /* CIRCLEQ_*, */
+#endif
+
 #include "build.h"
 
 /**
diff --git a/src/execve/enter.c b/src/execve/enter.c
index c562d14..814fc96 100644
--- a/src/execve/enter.c
+++ b/src/execve/enter.c
@@ -144,9 +144,11 @@ int translate_and_check_exec(Tracee *tracee, char host_path[PATH_MAX], const cha
 	if (status < 0)
 		return -ENOENT;
 
+#ifndef __OHOS__
 	status = access(host_path, X_OK);
 	if (status < 0)
 		return -EACCES;
+#endif
 
 	status = lstat(host_path, &statl);
 	if (status < 0)
diff --git a/src/extension/sysvipc/sysvipc_msg.c b/src/extension/sysvipc/sysvipc_msg.c
index 1fc290a..fd913b4 100644
--- a/src/extension/sysvipc/sysvipc_msg.c
+++ b/src/extension/sysvipc/sysvipc_msg.c
@@ -175,11 +175,17 @@ static int sysvipc_do_msgrcv(Tracee *tracee, struct SysVIpcConfig *config, size_
 		return -EINVAL;
 	}
 
+#ifdef MSG_COPY
 	if ((config->msgrcv_msgflg & ~(IPC_NOWAIT | MSG_NOERROR | MSG_COPY | MSG_EXCEPT)) != 0) {
 		return -EINVAL;
 	}
+#endif
 
+#ifdef MSG_COPY
 	bool copy = (config->msgrcv_msgflg & MSG_COPY) != 0;
+#else
+	bool copy = false;
+#endif
 	if (copy) {
 		if ((config->msgrcv_msgflg & IPC_NOWAIT) == 0) {
 			return -EINVAL;
diff --git a/src/path/binding.c b/src/path/binding.c
index c0e2f84..57a9c30 100644
--- a/src/path/binding.c
+++ b/src/path/binding.c
@@ -27,7 +27,11 @@
 #include <assert.h>   /* assert(3), */
 #include <limits.h>   /* PATH_MAX, */
 #include <errno.h>    /* E* */
+#ifdef __OHOS__
+#include "queue.h"    /* CIRCLEQ_*, */ // which is not present in ohos header
+#else
 #include <sys/queue.h> /* CIRCLEQ_*, */
+#endif
 #include <talloc.h>   /* talloc_*, */
 
 #include "path/binding.h"
diff --git a/src/path/binding.h b/src/path/binding.h
index b7f8d46..99c6e7e 100644
--- a/src/path/binding.h
+++ b/src/path/binding.h
@@ -26,6 +26,10 @@
 #include <limits.h> /* PATH_MAX, */
 #include <stdbool.h>
 
+#ifdef __OHOS__
+#include "queue.h"  /* CIRCLEQ_*,  */ // not present in ohos header
+#endif
+
 #include "tracee/tracee.h"
 #include "path.h"
 
diff --git a/src/path/path.c b/src/path/path.c
index d5c9ec6..04e4a3f 100644
--- a/src/path/path.c
+++ b/src/path/path.c
@@ -137,10 +137,12 @@ int which(Tracee *tracee, const char *paths, char host_path[PATH_MAX], const cha
 			return -EACCES;
 		}
 
+#ifndef __OHOS__
 		if (is_explicit && (statr.st_mode & S_IXUSR) == 0) {
 			note(tracee, ERROR, USER, "'%s' is not executable", command);
 			return -EACCES;
 		}
+#endif
 
 		found = true;
 
diff --git a/src/path/temp.c b/src/path/temp.c
index 8d8aa63..e726ec2 100644
--- a/src/path/temp.c
+++ b/src/path/temp.c
@@ -172,7 +172,7 @@ static int remove_temp_directory2(const char *path)
 	int status;
 	char *cwd;
 
-#ifdef __ANDROID__
+#if defined(__ANDROID__) || defined(__OHOS__)
 	cwd = malloc(PATH_MAX);
 	getcwd(cwd, PATH_MAX);
 #else
diff --git a/src/tracee/event.c b/src/tracee/event.c
index 756aa7e..158badd 100644
--- a/src/tracee/event.c
+++ b/src/tracee/event.c
@@ -80,10 +80,12 @@ int launch_process(Tracee *tracee, char *const argv[])
 		/* Declare myself as ptraceable before executing the
 		 * requested program. */
 		status = ptrace(PTRACE_TRACEME, 0, NULL, NULL);
+#ifndef __OHOS__
 		if (status < 0) {
 			note(tracee, ERROR, SYSTEM, "ptrace(TRACEME)");
 			return -errno;
 		}
+#endif
 
 		/* Synchronize with the tracer's event loop.  Without
 		 * this trick the tracer only sees the "return" from
diff --git a/src/tracee/statx.h b/src/tracee/statx.h
index d06dfae..f7dd286 100644
--- a/src/tracee/statx.h
+++ b/src/tracee/statx.h
@@ -5,6 +5,7 @@
 #include "sys/vfs.h"
 #include "path/path.h"
 #include "extension/extension.h"
+#include <linux/stat.h>
 
 /*
  * This structure is passed to extensions
diff --git a/src/tracee/tracee.c b/src/tracee/tracee.c
index a7082e3..9286d3f 100644
--- a/src/tracee/tracee.c
+++ b/src/tracee/tracee.c
@@ -26,7 +26,9 @@
 #include <assert.h>     /* assert(3), */
 #include <string.h>     /* bzero(3), */
 #include <stdbool.h>    /* bool, true, false, */
-#include <sys/queue.h>  /* LIST_*,  */
+#ifndef __OHOS__
+#include <sys/queue.h>  /* LIST_*,  */ // will interfere with added queue.h
+#endif
 #include <talloc.h>     /* talloc_*, */
 #include <signal.h>     /* kill(2), SIGKILL, */
 #include <sys/ptrace.h> /* ptrace(2), PTRACE_*, */
diff --git a/src/tracee/tracee.h b/src/tracee/tracee.h
index 3dcbc7b..4291b3f 100644
--- a/src/tracee/tracee.h
+++ b/src/tracee/tracee.h
@@ -26,7 +26,11 @@
 #include <sys/types.h> /* pid_t, size_t, */
 #include <sys/user.h>  /* struct user*, */
 #include <stdbool.h>   /* bool,  */
+#ifdef __OHOS__
+#include <queue.h>     /* CIRCLEQ_*, */ // which is not present in ohos header
+#else
 #include <sys/queue.h> /* LIST_*, */
+#endif
 #include <sys/ptrace.h>/* enum __ptrace_request */
 #include <talloc.h>    /* talloc_*, */
 #include <stdint.h>    /* *int*_t, */
