From f5fc915c332a2d02343d47866876a95b849d9e3e Mon Sep 17 00:00:00 2001
From: hackeris <hackeris@qq.com>
Date: Sat, 9 Aug 2025 00:33:28 +0800
Subject: [PATCH 3/5] update gitignore

---
 .gitignore | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/.gitignore b/.gitignore
index 3805615b8f4..68b76b9132e 100644
--- a/.gitignore
+++ b/.gitignore
@@ -1,7 +1,9 @@
 /GNUmakefile
 /build/
+/release/
 /.cache/
 /.vscode/
+/.idea/
 *.pyc
 .sdk
 .stgit-*
@@ -20,4 +22,3 @@ GTAGS
 *.swp
 *.patch
 *.gcov
-.idea
\ No newline at end of file
-- 
2.43.0

