From fac6d81ad6b2eafce20be33a9e02576201ea783c Mon Sep 17 00:00:00 2001
From: hackeris <hackeris@qq.com>
Date: Sat, 9 Aug 2025 00:53:24 +0800
Subject: [PATCH 4/5] enable jit on OHOS

---
 tcg/region.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/tcg/region.c b/tcg/region.c
index 7ea0b37a84c..718447044c2 100644
--- a/tcg/region.c
+++ b/tcg/region.c
@@ -572,6 +572,8 @@ static int alloc_code_gen_buffer_anon(size_t size, int prot,
 #ifndef CONFIG_TCG_INTERPRETER
 #ifdef CONFIG_POSIX
 #include "qemu/memfd.h"
+#include <sys/prctl.h>
+#define PRCTL_SET_JITFORT   0x6a6974
 
 static int alloc_code_gen_buffer_splitwx_memfd(size_t size, Error **errp)
 {
@@ -583,7 +585,14 @@ static int alloc_code_gen_buffer_splitwx_memfd(size_t size, Error **errp)
         goto fail;
     }
 
+    //  required by OHOS
+    prctl(PRCTL_SET_JITFORT, 0, 0);
+
     buf_rx = mmap(NULL, size, host_prot_read_exec(), MAP_SHARED, fd, 0);
+
+    //  required by OHOS
+    prctl(PRCTL_SET_JITFORT, 0, 1);
+
     if (buf_rx == MAP_FAILED) {
         error_setg_errno(errp, errno,
                          "failed to map shared memory for execute");
-- 
2.43.0

