From bdc5ec193a7366d5ba3a1afeb7fe5a9c847bf117 Mon Sep 17 00:00:00 2001
From: hackeris <hackeris@qq.com>
Date: Sun, 6 Jul 2025 01:40:32 +0800
Subject: [PATCH 4/5] Drop PROT_EXEC for mmap on OHOS

---
 linux-user/syscall.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/linux-user/syscall.c b/linux-user/syscall.c
index c7e8425..0e9cc55 100644
--- a/linux-user/syscall.c
+++ b/linux-user/syscall.c
@@ -9119,10 +9119,13 @@ static abi_long do_syscall1(void *cpu_env, int num, abi_long arg1,
                                         v5, v6));
         }
 #else
-        ret = get_errno(target_mmap(arg1, arg2, arg3,
+        {
+            int prot = arg3 & (~PROT_EXEC);
+            ret = get_errno(target_mmap(arg1, arg2, prot,
                                     target_to_host_bitmask(arg4, mmap_flags_tbl),
                                     arg5,
                                     arg6));
+        }
 #endif
         return ret;
 #endif
-- 
2.49.0.windows.1

