From 957e419e1652415cf442f306c0ec819116aae065 Mon Sep 17 00:00:00 2001
From: hackeris <hackeris@qq.com>
Date: Sat, 9 Aug 2025 10:54:00 +0800
Subject: [PATCH 5/5] fix argv for geust on OHOS

---
 linux-user/main.c | 7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

diff --git a/linux-user/main.c b/linux-user/main.c
index 68972f00a15..3f8335d6c8e 100644
--- a/linux-user/main.c
+++ b/linux-user/main.c
@@ -804,9 +804,10 @@ int main(int argc, char **argv, char **envp)
      *    argv[optind]     full path to the binary
      *    argv[optind + 1] original argv[0]
      */
-    if (optind + 1 < argc && preserve_argv0) {
-        optind++;
-    }
+    //  disabled for OHOS
+    // if (optind + 1 < argc && preserve_argv0) {
+    //     optind++;
+    // }
 
     if (cpu_model == NULL) {
         cpu_model = cpu_get_model(get_elf_eflags(execfd));
-- 
2.43.0

