From 589fd77129b9bada6bcbe4a2dac064e19b3c2468 Mon Sep 17 00:00:00 2001
From: hackeris <hackeris@qq.com>
Date: Sat, 23 Aug 2025 20:38:15 +0800
Subject: [PATCH 7/9] fix for x86 target

---
 tcg/i386/tcg-target.c.inc | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/tcg/i386/tcg-target.c.inc b/tcg/i386/tcg-target.c.inc
index 088c6c9264b..240ac15ad5a 100644
--- a/tcg/i386/tcg-target.c.inc
+++ b/tcg/i386/tcg-target.c.inc
@@ -2123,7 +2123,10 @@ static HostAddress x86_guest_base = {
 #if defined(__x86_64__) && defined(__linux__)
 # include <asm/prctl.h>
 # include <sys/prctl.h>
-int arch_prctl(int code, unsigned long addr);
+int arch_prctl(int code, unsigned long addr)
+{
+    return syscall(SYS_arch_prctl, code, addr);
+}
 static inline int setup_guest_base_seg(void)
 {
     if (arch_prctl(ARCH_SET_GS, guest_base) == 0) {
-- 
2.43.0

