From 09a1185af81e7180075d2cfee25ab0871d57644e Mon Sep 17 00:00:00 2001
From: hackeris <hackeris@qq.com>
Date: Sat, 9 Aug 2025 00:29:53 +0800
Subject: [PATCH 2/6] fix pc register from tci to guest

---
 linux-user/signal.c | 6 ++++++
 tcg/tci.c           | 3 +++
 2 files changed, 9 insertions(+)

diff --git a/linux-user/signal.c b/linux-user/signal.c
index daaa2ef7489..816d4872a7b 100644
--- a/linux-user/signal.c
+++ b/linux-user/signal.c
@@ -37,6 +37,8 @@
 #include "user/page-protection.h"
 #include "user/safe-syscall.h"
 #include "user/signal.h"
+
+#include "accel/tcg/getpc.h"
 #include "tcg/tcg.h"
 
 /* target_siginfo_t must fit in gdbstub's siginfo save area. */
@@ -966,7 +968,11 @@ static void host_sigsegv_handler(CPUState *cpu, siginfo_t *info,
      */
     bool is_valid = h2g_valid(host_addr);
     abi_ptr guest_addr = h2g_nocheck(host_addr);
+#ifndef CONFIG_TCG_INTERPRETER
     uintptr_t pc = host_signal_pc(uc);
+#else
+    uintptr_t pc = GETPC();
+#endif
     bool is_write = host_signal_write(info, uc);
     MMUAccessType access_type = adjust_signal_pc(&pc, is_write);
     bool maperr;
diff --git a/tcg/tci.c b/tcg/tci.c
index 700e6726169..59bfd11e803 100644
--- a/tcg/tci.c
+++ b/tcg/tci.c
@@ -357,6 +357,9 @@ uintptr_t QEMU_DISABLE_CFI tcg_qemu_tb_exec(CPUArchState *env,
         int32_t ofs;
         void *ptr;
 
+#ifdef CONFIG_TCG_INTERPRETER
+        tci_tb_ptr = (uintptr_t)tb_ptr;
+#endif
         insn = *tb_ptr++;
         opc = extract32(insn, 0, 8);
 
-- 
2.43.0

