diff --git a/linux-user/syscall.c b/linux-user/syscall.c
index 19c5dc7..3a376f1 100644
--- a/linux-user/syscall.c
+++ b/linux-user/syscall.c
@@ -9116,10 +9116,13 @@ static abi_long do_syscall1(void *cpu_env, int num, abi_long arg1,
                                         v5, v6));
         }
 #else
-        ret = get_errno(target_mmap(arg1, arg2, arg3,
+        {
+            int prot = arg3 & (~PROT_EXEC);
+            ret = get_errno(target_mmap(arg1, arg2, prot,
                                     target_to_host_bitmask(arg4, mmap_flags_tbl),
                                     arg5,
                                     arg6));
+        }
 #endif
         return ret;
 #endif
