From 2220086be06de6da0089059dce2e8f2bfe2747ea Mon Sep 17 00:00:00 2001
From: Jiajie Chen <c@jia.je>
Date: Sun, 6 Jul 2025 09:24:03 +0800
Subject: [PATCH 6/7] Disable setuid/setgid on OHOS

---
 linux-user/syscall.c | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/linux-user/syscall.c b/linux-user/syscall.c
index b1f1988c92..f52e9cd391 100644
--- a/linux-user/syscall.c
+++ b/linux-user/syscall.c
@@ -10949,9 +10949,13 @@ static abi_long do_syscall1(void *cpu_env, int num, abi_long arg1,
         return ret;
 #endif
     case TARGET_NR_setuid:
-        return get_errno(sys_setuid(low2highuid(arg1)));
+        // we cannot use setuid on OHOS
+        // return get_errno(sys_setuid(low2highuid(arg1)));
+        return 0;
     case TARGET_NR_setgid:
-        return get_errno(sys_setgid(low2highgid(arg1)));
+        // we cannot use setgid on OHOS
+        // return get_errno(sys_setgid(low2highgid(arg1)));
+        return 0;
     case TARGET_NR_setfsuid:
         return get_errno(setfsuid(arg1));
     case TARGET_NR_setfsgid:
-- 
2.49.0

