From bf32981f17eff55c3e517b073fd072cc99b98bc4 Mon Sep 17 00:00:00 2001
From: Jiajie Chen <c@jia.je>
Date: Sun, 6 Jul 2025 09:50:34 +0800
Subject: [PATCH 7/7] Find qemu abs path from /proc/self/exe instead

---
 linux-user/execve.c | 13 ++++---------
 1 file changed, 4 insertions(+), 9 deletions(-)

diff --git a/linux-user/execve.c b/linux-user/execve.c
index 3168b85e79..349267b257 100644
--- a/linux-user/execve.c
+++ b/linux-user/execve.c
@@ -2,6 +2,7 @@
 
 #include <stdlib.h>
 #include <string.h>
+#include <assert.h>
 #include <unistd.h>
 #include <linux/limits.h>
 
@@ -10,15 +11,9 @@ char qemu_abs_path[PATH_MAX];
 const char **argv_prefix;
 
 void setup_for_execve(const char **argv, int optind) {
-
-    const char *qemu = argv[0];
-    if (qemu[0] != '/') {
-        getcwd(qemu_abs_path, PATH_MAX);
-        strcat(qemu_abs_path, "/");
-        strcat(qemu_abs_path, qemu);
-    } else {
-        strcat(qemu_abs_path, qemu);
-    }
+    ssize_t ret = readlink("/proc/self/exe", qemu_abs_path, PATH_MAX - 1);
+    assert(ret > 0);
+    qemu_abs_path[ret] = '\0';
 
     argv_prefix = calloc(optind + 1, sizeof(char *));
     argv_prefix[0] = qemu_abs_path;
-- 
2.49.0

