From 8dbd5728dfb26ffde96237f8938f8506ea5d6d48 Mon Sep 17 00:00:00 2001
From: hackeris <hackeris@qq.com>
Date: Thu, 14 Aug 2025 00:06:09 +0800
Subject: [PATCH 8/8] forward /dev access to host

---
 util/path.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/util/path.c b/util/path.c
index f602d351072..a2dcc178476 100644
--- a/util/path.c
+++ b/util/path.c
@@ -48,7 +48,9 @@ static bool skip_relocation(const char* name)
         || strcmp(name, "/sys") == 0
         || strcmp(name, "/etc/resolv.conf") == 0
         || strcmp(name, "/etc/passwd") == 0
-        || strcmp(name, "/dev/null") == 0;
+        || strcmp(name, "/dev") == 0
+        || strstr(name, "/dev/") == name
+    ;
 }
 
 const char* do_relocate_path(const char* name, char* out)
-- 
2.43.0

