From 58dd0d4d36d599ab7ce8088cd2e07e57e7566325 Mon Sep 17 00:00:00 2001
From: hackeris <hackeris@qq.com>
Date: Fri, 18 Jul 2025 10:06:41 +0800
Subject: [PATCH 22/22] Replace renameat on symlink by symlinkat and unlinkat

---
 linux-user/syscall.c | 28 +++++++++++++++++++++++++++-
 1 file changed, 27 insertions(+), 1 deletion(-)

diff --git a/linux-user/syscall.c b/linux-user/syscall.c
index 366c639..4710c8c 100644
--- a/linux-user/syscall.c
+++ b/linux-user/syscall.c
@@ -8333,8 +8333,34 @@ static abi_long do_syscall1(void *cpu_env, int num, abi_long arg1,
             else {
                 char reloc1[PATH_MAX];
                 char reloc2[PATH_MAX];
-                ret = get_errno(renameat(arg1, relocate_path_at(arg1, p, reloc1, false), arg3, relocate_path_at(arg3, p2, reloc2, false)));
+                relocate_path_at(arg1, p, reloc1, false);
+                relocate_path_at(arg3, p2, reloc2, false);
+                ret = get_errno(renameat(arg1, reloc1, arg3, reloc2));
+                if (ret < 0 && errno == EACCES) {
+                    struct stat s;
+                    if (lstat(reloc1, &s) < 0) {
+                        goto renameat_end;
+                    }
+                    if (!S_ISLNK(s.st_mode)) {
+                        goto renameat_end;
+                    }
+                    char link[PATH_MAX];
+                    ssize_t sz = readlinkat(arg1, reloc1, link, PATH_MAX - 1);
+                    if (sz < 0) {
+                        goto renameat_end;
+                    }
+                    link[sz] = '\0';
+                    if (symlinkat(link, arg3, reloc2) < 0) {
+                        goto renameat_end;
+                    }
+                    if (unlinkat(arg1, reloc1, 0) < 0) {
+                        goto renameat_end;
+                    }
+                    ret = 0;
+                    errno = 0;
+                }
             }
+        renameat_end:
             unlock_user(p2, arg4, 0);
             unlock_user(p, arg2, 0);
         }
-- 
2.49.0.windows.1

