From 6b52ada67786c89a23c0ee771cad25bee7a3ba57 Mon Sep 17 00:00:00 2001
From: hackeris <hackeris@qq.com>
Date: Tue, 22 Jul 2025 00:11:10 +0800
Subject: [PATCH 29/32] Fix NOFOLLOW flags for NR_statx

---
 linux-user/syscall.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/linux-user/syscall.c b/linux-user/syscall.c
index 3d6c25b..b79336e 100644
--- a/linux-user/syscall.c
+++ b/linux-user/syscall.c
@@ -10997,7 +10997,7 @@ static abi_long do_syscall1(void *cpu_env, int num, abi_long arg1,
                 struct target_statx host_stx;
                 int mask = arg4;
 
-                bool follow_symlink = arg4 & AT_SYMLINK_NOFOLLOW ? false : true;
+                bool follow_symlink = flags & AT_SYMLINK_NOFOLLOW ? false : true;
                 char reloc[PATH_MAX];
                 ret = get_errno(sys_statx(dirfd, relocate_path_at(dirfd, p, reloc, follow_symlink), flags, mask, &host_stx));
                 if (!is_error(ret)) {
@@ -11014,7 +11014,7 @@ static abi_long do_syscall1(void *cpu_env, int num, abi_long arg1,
             }
 #endif
             {
-                bool follow_symlink = arg4 & AT_SYMLINK_NOFOLLOW ? false : true;
+                bool follow_symlink = flags & AT_SYMLINK_NOFOLLOW ? false : true;
                 char reloc[PATH_MAX];
                 ret = get_errno(fstatat(dirfd, relocate_path_at(dirfd, p, reloc, follow_symlink), &st, flags));
                 unlock_user(p, arg2, 0);
-- 
2.49.0.windows.1

