From 917db534e9c23ffaf25f174613ffcf0132af487f Mon Sep 17 00:00:00 2001
From: hackeris <hackeris@qq.com>
Date: Tue, 22 Jul 2025 00:46:50 +0800
Subject: [PATCH 30/32] Reuse passwd file of host

---
 util/path.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/util/path.c b/util/path.c
index b18a1e6..e23eccf 100644
--- a/util/path.c
+++ b/util/path.c
@@ -43,6 +43,7 @@ static bool skip_relocation(const char *name) {
 	return strstr(name, "/proc/") == name
 	       || strcmp(name, "/proc") == 0
 	       || strcmp(name, "/etc/resolv.conf") == 0
+	       || strcmp(name, "/etc/passwd") == 0
 	       || strcmp(name, "/dev/null") == 0
 	;
 }
-- 
2.49.0.windows.1

