--- a/linux-user/aarch64/syscall_64.tbl	2025-05-29 06:05:46.000000000 +0800
+++ b/linux-user/aarch64/syscall_64.tbl	2025-06-23 00:46:33.450846576 +0800
@@ -218,14 +218,6 @@
 177	common	getegid				sys_getegid
 178	common	gettid				sys_gettid
 179	common	sysinfo				sys_sysinfo			compat_sys_sysinfo
-180	common	mq_open				sys_mq_open			compat_sys_mq_open
-181	common	mq_unlink			sys_mq_unlink
-182	time32	mq_timedsend			sys_mq_timedsend_time32
-182	64	mq_timedsend			sys_mq_timedsend
-183	time32	mq_timedreceive			sys_mq_timedreceive_time32
-183	64	mq_timedreceive			sys_mq_timedreceive
-184	common	mq_notify			sys_mq_notify			compat_sys_mq_notify
-185	common	mq_getsetattr			sys_mq_getsetattr		compat_sys_mq_getsetattr
 186	common	msgget				sys_msgget
 187	common	msgctl				sys_msgctl			compat_sys_msgctl
 188	common	msgrcv				sys_msgrcv			compat_sys_msgrcv
