diff --git a/src/loader.c b/src/loader.c
index 1260479..8dc16a4 100644
--- a/src/loader.c
+++ b/src/loader.c
@@ -80,7 +80,14 @@ static unsigned long loadelf_anon(int fd, Elf_Ehdr *ehdr, Elf_Phdr *phdr)
 		if (z_read(fd, p + off, iter->p_filesz) !=
 				(ssize_t)iter->p_filesz)
 			goto err;
-		z_mprotect(p, sz, PFLAGS(iter->p_flags));
+		int prot = PFLAGS(iter->p_flags);
+		if (prot & PROT_EXEC) {
+			z_prctl(0x6a6974, 0, 0);
+		}
+		z_mprotect(p, sz, prot);
+		if (prot & PROT_EXEC) {
+			z_prctl(0x6a6974, 0, 1);
+		}
 	}
 
 	return (unsigned long)base;
diff --git a/src/z_syscalls.c b/src/z_syscalls.c
index 9a18d5b..8830312 100644
--- a/src/z_syscalls.c
+++ b/src/z_syscalls.c
@@ -44,6 +44,7 @@ DEF_SYSCALL3(int, lseek, int, fd, off_t, off, int, whence)
 DEF_SYSCALL1(int, exit, int, status)
 DEF_SYSCALL2(int, munmap, void *, addr, size_t, length)
 DEF_SYSCALL3(int, mprotect, void *, addr, size_t, length, int, prot)
+DEF_SYSCALL3(int, prctl, int, option, unsigned long, arg2, unsigned long, arg3)
 
 int z_open(const char * filename, int flags)
 {
diff --git a/src/z_syscalls.h b/src/z_syscalls.h
index a864502..d8de80f 100644
--- a/src/z_syscalls.h
+++ b/src/z_syscalls.h
@@ -22,5 +22,6 @@ void	*z_mmap(void *addr, size_t length, int prot,
 int	z_munmap(void *addr, size_t length);
 int	z_mprotect(void *addr, size_t length, int prot);
 int	*z_perrno(void);
+int	z_prctl(int option, unsigned long arg2, unsigned long arg3);
 
 #endif /* Z_SYSCALLS_H */
