--- a/gdbserver/linux-amd64-ipa.cc	2025-04-21 01:22:06.000000000 +0800
+++ b/gdbserver/linux-amd64-ipa.cc	2025-08-28 00:04:25.067615107 +0800
@@ -233,7 +233,7 @@
   return NULL;
 #else
   void *res = mmap (NULL, size, PROT_READ | PROT_WRITE | PROT_EXEC,
-		    MAP_PRIVATE | MAP_ANONYMOUS | MAP_32BIT, -1, 0);
+		    MAP_PRIVATE | MAP_ANONYMOUS , -1, 0);
 
   if (res == MAP_FAILED)
     return NULL;
