--- a/ssh.c	2025-06-09 19:34:19
+++ b/ssh.c	2025-06-09 19:34:23
@@ -713,7 +713,7 @@
 	seed_rng();
 
 	/* Get user data. */
-	pw = getpwuid(getuid());
+	pw = mygetpwuid(getuid());
 	if (!pw) {
 		logit("No user exists for uid %lu", (u_long)getuid());
 		exit(255);
--- a/misc.c	2025-06-09 19:33:27
+++ b/misc.c	2025-06-09 19:34:29
@@ -1236,7 +1236,7 @@
 			error_f("No such user %s", user);
 			goto out;
 		}
-	} else if ((pw = getpwuid(uid)) == NULL) {
+	} else if ((pw = mygetpwuid(uid)) == NULL) {
 		error_f("No such uid %ld", (long)uid);
 		goto out;
 	}
@@ -3137,4 +3137,16 @@
 		return 1;
 	}
 	return 0;
+}
+
+struct passwd *mygetpwuid(uid_t uid) {
+	static struct passwd pw;
+	pw.pw_name = "currentUser";
+	pw.pw_passwd = NULL;
+	pw.pw_gecos = "";
+	pw.pw_uid = getuid();
+	pw.pw_gid = getuid();
+	pw.pw_dir = "/storage/Users/currentUser";
+	pw.pw_shell = "/bin/sh";
+	return &pw;
 }
--- a/misc.h	2025-06-09 19:32:45
+++ b/misc.h	2025-06-09 19:33:22
@@ -257,5 +257,7 @@
 
 /* On OpenBSD time_t is int64_t which is long long. */
 /* #define SSH_TIME_T_MAX LLONG_MAX */
+
+struct passwd *mygetpwuid(uid_t uid);
 
 #endif /* _MISC_H */
