--- vim-9.1.1436/src/viminfo.c	2025-06-07 23:49:35
+++ ../viminfo.c	2025-06-11 12:46:27
@@ -13,6 +13,7 @@
 
 #include "vim.h"
 #include "version.h"
+#include <unistd.h>
 
 /*
  * Structure used for reading from the viminfo file.
@@ -3087,6 +3088,24 @@
     return OK;
 }
 
+int gid_match(gid_t gid) {
+	if (gid == getgid()) {
+		return OK;
+	}
+	int size = getgroups(0, NULL);
+	gid_t *groups = (gid_t *)malloc(sizeof(gid_t) * size);
+	assert(groups);
+	getgroups(size, groups);
+	for (int i = 0;i < size;i++) {
+		if (groups[i] == gid) {
+			free(groups);
+			return OK;
+		}
+	}
+	free(groups);
+	return FAIL;
+}
+
 /*
  * Write the viminfo file.  The old one is read in first so that effectively a
  * merge of current info and old info is done.  This allows multiple vims to
@@ -3154,7 +3173,7 @@
 		|| (getuid() != ROOT_UID
 		    && !(st_old.st_uid == getuid()
 			    ? (st_old.st_mode & 0200)
-			    : (st_old.st_gid == getgid()
+			    : (gid_match(st_old.st_gid)
 				    ? (st_old.st_mode & 0020)
 				    : (st_old.st_mode & 0002))))
 #endif
